### DISCUSSION TITLE
New to GLSL: blending diffuse color ramp with color from lights?

### DISCUSSION THREAD

**Discussion by Klimt1234** [bnpr/Malt#353] [Q&A]
Hi,
I've been trying to make a toon shader in Evee that can blend the diffuse color ramp gradient with colors from indirect lighting. I came across Malt in my experimentation and looked into GLSL coding tutorials for several weeks now but I can't make any sense of it, as I've never touched a line of code before.

Is my objective even possible or do I just want too much from life? If it's possible, can anyone please explain in simple terms how that can be achieved? It would be much appreciated.

Thanks

**Comment by pragma37** [comment]
Sorry, I'm not sure what do you mean exactly.

Could you provide more details about what is blending with the colors from indirect lighting? And what indirect lighting means in the context of Malt?

You don't need to touch GLSL to use Malt, btw. Most things can be done with nodes.

**Reply by Klimt1234** [reply]
Hi,
What I mean by 'color from indirect lighting' is the various colors that can be observed in an environment from light rays bouncing off various surfaces and 'carrying' some of those colors with these bounces. 

For example, if you put a white sphere on a blue floor, then shine a light on the sphere, you'll notice that the sphere has a blue tint on the area where the light bounces between the sphere and the floor. That blue tint will be a mix of the blue from the floor, white from the sphere, and whatever color the light is. 

Another example would be Impressionist or Fauvist paintings. They depict their subjects in a variety of colors that would naturally be cast by the lighting of the enviorment they are painted in. This gives the painted subjects and the composition more depth and more believability.

That is essentially what I would like this shader to do. I want it to display it's own color gradient, as well as mix with the light colors in an enviorment. So far in Eevee, I've only been able to do one or the other, not both. That is why I ask in this forum to see if Malt can produce this kind of result.

Also, if I understand correctly, don't Malt nodes represent GLSL code?

Thanks

**Reply by pragma37** [reply]
The problem is that Malt doesn't support indirect lighting yet (neither EEVEE, unless you are using irradiance volumes?).

I think your best option at the moment would be to do the diffuse gradient part in either Malt or EEVEE and then use compositing to blend it with a Cycles Diffuse Indirect pass.

> Also, if I understand correctly, don't Malt nodes represent GLSL code?

Yes, but you can use them without understanding GLSL. (EEVEE nodes are compiled into GLSL too)

**Reply by shankarsivarajan** [reply]
Is there a way to export the GLSL code that Malt generates? That might prove useful for game engines.
